### PR TITLE
AH-436: update library

### PR DESCRIPTION
**Pull request by jeanmurillo95**
# Surf-shop Change Request

## Change Comments

I updated the mongodb library, I also changed some methods and test cases. 

## Solution description

I updated the mongodb package, I reduced the complexity in some functions, repair some test cases and lint errors.

## Linting & unit tests ev

Screen Shot 2022-08-30 at 12 45 59 PM
idence

## Checklist

_**PR is ready for code review and approval when each box is checked.**_

_All steps are required, when applicable. ~Strikeout~ formatting indicates a step is not applicable to this change set._

- [x] Ticket referenced in PR title (ex. _AH-XXX: Short summary of change_)
- [x] Not-applicable items in this checklist struck out

### CODE REVIEW

**Review by candradeg — APPROVED** [review]
lgtm

**Review by Fernavarro21 — APPROVED** [review]
lgtm

**Review by daniel-hume — APPROVED** [review]
looks good

File: yarn.lock [hunk 1]
```diff
@@ -0,0 +1,3789 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
```
Comment (DanStory) [inline yarn.lock]: What's the reasoning we are switching package management to yarn?

File: tests/Repository.spec.ts [hunk 2]
```diff
@@ -80,7 +80,7 @@ describe('MongoRepository', () => {
         if (!this.events[eventArgs.operationType][eventArgs.operation]) {
           this.events[eventArgs.operationType][eventArgs.operation] = [];
         }
-        this.events[eventArgs.operationType][eventArgs.operation].push({ args: clone(args) });
+        this.events[eventArgs.operationType][eventArgs.operation].push({ args: _.cloneDeep(args) });
```
Comment (DanStory) [inline tests/Repository.spec.ts]: `rfdc` already solves this, seems like bloat to be using `lodash` solely for this?

File: package.json [hunk 3]
```diff
@@ -36,32 +36,33 @@
...
     "tslint-config-swimlane": "^5.0.0",
-    "typescript": "^4.1.5"
+    "typescript": "^4.7.4"
   },
   "dependencies": {
-    "mongodb": "3.6.6",
-    "retry": "^0.12.0",
-    "rfdc": "^1.2.0"
+    "lodash": "^4.17.21",
+    "mongodb": "^4.9.0",
```
Comment (DanStory) [inline package.json]: this mongo upgrade cases breaking changes to the repo signatures, we need to bump the package.json major version as well before publish

File: package.json [hunk 4]
```diff
@@ -36,32 +36,33 @@
...
-    "codacy-coverage": "^3.2.0",
-    "faker": "^4.1.0",
-    "mocha": "^9.1.3",
-    "mongo-mock": "^4.0.0",
-    "nyc": "^15.0.0",
-    "prettier": "~2.2.1",
+    "@types/chai": "^4.3.3",
+    "@types/faker": "^6.6.9",
+    "@types/mocha": "^9.1.1",
+    "@types/mongodb": "^4.0.7",
```
Comment (DanStory) [inline package.json]: according to npm, [`mongodb`](https://www.npmjs.com/package/mongodb)has the types built-in and [`@types/mongodb`](https://www.npmjs.com/package/@types/mongodb) is deprecated

File: package.json [hunk 5]
```diff
@@ -36,32 +36,33 @@
...
+    "@types/node": "^18.7.13",
+    "@types/retry": "^0.12.2",
+    "chai": "^4.3.6",
+    "codacy-coverage": "^3.4.0",
+    "faker": "^5.5.3",
```
Comment (DanStory) [inline package.json]: We should probably get off this for `@faker-js/faker`, due to the demise of the original `faker`

File: package.json [hunk 6]
```diff
@@ -36,32 +36,33 @@
...
-    "chai": "^4.2.0",
-    "clone": "^2.1.2",
-    "codacy-coverage": "^3.2.0",
-    "faker": "^4.1.0",
-    "mocha": "^9.1.3",
-    "mongo-mock": "^4.0.0",
-    "nyc": "^15.0.0",
-    "prettier": "~2.2.1",
+    "@types/chai": "^4.3.3",
+    "@types/faker": "^6.6.9",
```
Comment (DanStory) [inline package.json]: This needs to be the same major.minor version of `faker` if we are gonna stay on it for now. Anything on v6 can't be trusted.

File: package.json [hunk 7]
```diff
@@ -36,32 +36,33 @@
...
+    "@types/mocha": "^9.1.1",
+    "@types/mongodb": "^4.0.7",
+    "@types/node": "^18.7.13",
+    "@types/retry": "^0.12.2",
+    "chai": "^4.3.6",
+    "codacy-coverage": "^3.4.0",
+    "faker": "^5.5.3",
+    "jest": "^29.0.0",
```
Comment (DanStory) [inline package.json:47]: We should add `@types/jest`

File: package.json [hunk 8]
```diff
@@ -36,32 +36,33 @@
...
-    "faker": "^4.1.0",
-    "mocha": "^9.1.3",
-    "mongo-mock": "^4.0.0",
-    "nyc": "^15.0.0",
-    "prettier": "~2.2.1",
+    "@types/chai": "^4.3.3",
+    "@types/faker": "^6.6.9",
+    "@types/mocha": "^9.1.1",
+    "@types/mongodb": "^4.0.7",
+    "@types/node": "^18.7.13",
```
Comment (DanStory) [inline package.json]: lets stay on LTS (`v16`)

File: src/Repository.ts [hunk 9]
```diff
@@ -46,8 +47,8 @@ export class MongoRepository<DOC, DTO = DOC> {
    * @returns {Promise<DOC | null>}
    * @memberof MongoRepository
    */
-  findById(id: string): Promise<DOC | null> {
-    return this.findOne({ _id: new ObjectID(id) });
+  findById(id: string): Promise<WithId<DOC> | null> {
```
Comment (DanStory) [inline src/Repository.ts:50]: `WithId` is for the Mongo Document `_Id` field, the MongoRepoisotry transforms them to `Id` via `toggleId`, so this would be contradicting the actual result.

File: src/Repository.ts [hunk 10]
```diff
@@ -288,15 +289,23 @@ export class MongoRepository<DOC, DTO = DOC> {
...
   protected toggleId(document: any, replace: boolean): any {
     if (document && (document.id || document._id)) {
-      if (replace) {
-        document._id = new ObjectID(document.id);
+      this.setId(document, replace);
+    }
+    return document;
+  }
+
+  protected setId(document: any, replace: boolean): void {
```
Comment (DanStory) [inline src/Repository.ts]: There doesn't seem to be any value of refactoring this out to another method.

File: src/Repository.ts [hunk 11]
```diff
@@ -322,30 +331,46 @@ export class MongoRepository<DOC, DTO = DOC> {
...
-          newDocument = await newDocument;
-        }
-      }
+      await this.bindEvents(_.cloneDeep(events), fn, type, originalDocument, newDocument);
     }
-
     return newDocument;
   }
 
+  protected cloneDocuments(newDocument: any, originalDocument: any): void {
```
Comment (DanStory) [inline src/Repository.ts:345]: This won't work as intended, the caller will never get the new cloned instances and would still be operating on the original object instances.

File: src/Repository.ts [hunk 12]
```diff
@@ -322,30 +331,46 @@ export class MongoRepository<DOC, DTO = DOC> {
...
+    newDocument = _.cloneDeep(newDocument);
+    if (originalDocument) {
+      originalDocument = _.cloneDeep(originalDocument);
+    }
+  }
+
+  protected async bindEvents(
```
Comment (DanStory) [inline src/Repository.ts]: This refactor seems less intuitive, I don't see value in going from a for...loop to a recursion that also requires cloning the arguments for manipulation.

File: src/Repository.ts [hunk 13]
```diff
@@ -357,54 +382,65 @@ export class MongoRepository<DOC, DTO = DOC> {
...
+    let indexDefinition;
+    try {
+      if (indexesOptions?.length > 0) {
+        indexDefinition = indexesOptions.shift();
+        await ourCollection.createIndex(indexDefinition.fields, indexDefinition.options);
+        return this.indexDefinition(ourCollection, indexesOptions);
+      }
+      return;
+    } catch (indexErr) {
+      this.reTryIndexCreation(indexDefinition, ourCollection, indexesOptions, indexErr);
```
Comment (DanStory) [inline src/Repository.ts]: This sub-recursion refactor is hard to follow and the argument manipulation (array.shift) can be unexpected from a caller's perspective.

File: src/Repository.ts [hunk 14]
```diff
@@ -357,54 +382,65 @@ export class MongoRepository<DOC, DTO = DOC> {
...
+
+  private async createCollection(db: Db): Promise<Collection<DOC> | Collection<Document>> {
+    try {
+      return await db.createCollection(this.options.name, {
+        size: this.options.size,
+        capped: this.options.capped,
+        max: this.options.max
+      });
+    } catch (createErr) {
+      if (createErr.codeName === 'NamespaceExists') {
```
Comment (DanStory) [inline src/Repository.ts]: We lost comment context in this refactor

File: src/Repository.ts [hunk 15]
```diff
@@ -357,54 +382,65 @@ export class MongoRepository<DOC, DTO = DOC> {
...
+        await ourCollection.createIndex(indexDefinition.fields, indexDefinition.options);
+        return this.indexDefinition(ourCollection, indexesOptions);
+      }
+      return;
+    } catch (indexErr) {
+      this.reTryIndexCreation(indexDefinition, ourCollection, indexesOptions, indexErr);
+    }
+  }
+
+  private async reTryIndexCreation(
```
Comment (DanStory) [inline src/Repository.ts]: retry is one word, should be lowercase (`retryIndexCreation`)

File: src/Repository.ts [hunk 16]
```diff
@@ -357,54 +382,65 @@ export class MongoRepository<DOC, DTO = DOC> {
...
+    ) {
+      try {
+        await ourCollection.dropIndex(indexDefinition.options.name);
+        return await this.indexDefinition(ourCollection, indexesOptions);
+      } catch (recreateErr) {
+        throw recreateErr;
+      }
+    }
+    throw indexErr;
+  }
```
Comment (DanStory) [inline src/Repository.ts]: Be sure to add docblocks (at least summaries) to the functions

**Review by DanStory — CHANGES_REQUESTED** [review]

File: tests/Repository.spec.ts [hunk 17]
```diff
@@ -5,7 +5,7 @@ mongoMock.max_delay = 0; // turn of fake async
 import { expect } from 'chai';
 import * as faker from 'faker';
 import { ObjectId } from 'mongodb';
-import * as clone from 'clone';
+import * as _ from 'lodash';
```
Comment (DanStory) [inline tests/Repository.spec.ts]: Should be replaced with `rfdc`

File: yarn.lock (hunk 1 above)
Comment (jeanmurillo95) [reply yarn.lock]: @DanStory When I was downloading the package using npm the second time the npm install failed because the package-lock.json had some errors so I preferred to use yarn to avoid this problem.

File: tests/Repository.spec.ts (hunk 2 above)
Comment (jeanmurillo95) [reply tests/Repository.spec.ts]: @DanStory  no, rdfc has trouble cloning the new ObjectId type, so if you need this ObjectId, the returned value has strange behavior.

File: package.json (hunk 3 above)
Comment (jeanmurillo95) [reply package.json]: What do you refer about the repo signatures?

File: package.json (hunk 4 above)
Comment (jeanmurillo95) [reply package.json]: You’re right, mongodb provides its own types, I will delete this dependency

File: package.json (hunk 5 above)
Comment (jeanmurillo95) [reply package.json]: Why do you prefer this package?

File: src/Repository.ts (hunk 16 above)
Comment (jeanmurillo95) [reply src/Repository.ts]: Sure, I will do.

File: src/Repository.ts (hunk 13 above)
Comment (jeanmurillo95) [reply src/Repository.ts]: The previous function was not fulfilling the solid principles, so I thought to separate it by feature, what do you recommend me?

File: src/Repository.ts (hunk 14 above)
Comment (jeanmurillo95) [reply src/Repository.ts]: I'll solve it

File: src/Repository.ts (hunk 15 above)
Comment (jeanmurillo95) [reply src/Repository.ts]: I'll solve it

File: src/Repository.ts (hunk 12 above)
Comment (jeanmurillo95) [reply src/Repository.ts]: I using this cloning because If I don’t used It some test cases will fail. and I changed this for to a recursive because I'm trying to avoid using a foor loop inside of other, and the recursion is more efficient.

Comment (DanStory) [reply src/Repository.ts]: Array.pop and recursion is going to be slower than a for.loop in general. Because you are modifying the argument array instance (`events.pop()`), this is modifying the same instance up the call stack, creating unintended side-effects. Most likely the cause for tests to the fail. Using a for.loop would not require the array manipulation, recursion, or cloning.

I personally don't see an issue with a second deep for.loop but you can keep it as a separate method if you like, I however would suggest this is a `private` method at this point.

File: package.json (hunk 5 above)
Comment (DanStory) [reply package.json]: [What happened to `faker`](https://fakerjs.dev/about/announcements/2022-01-14.html#i-heard-something-happened-what-s-the-tldr)

`faker-js/faker` is the "new-owners" community driven of the original `faker`

File: yarn.lock (hunk 1 above)
Comment (DanStory) [reply yarn.lock]: I would suggest we fix the npm package-lock errors vs introducing a different package management that could require  additional build pipeline and developer environment changes.

Simplest resolution would be to delete the `package-lock.json` and run `npm install` to regen it.

File: package.json (hunk 3 above)
Comment (DanStory) [reply package.json]: The mongo types MongoRepository exposes have changed, i.e. `DeleteWriteOpResultObject`->`DeleteResult`, possibly `UpdateRequest` properties, etc.

File: src/Repository.ts [hunk 18]
```diff
@@ -239,8 +240,8 @@ export class MongoRepository<DOC, DTO = DOC> {
    * @returns {Promise<DeleteWriteOpResultObject>}
```
Comment (DanStory) [inline src/Repository.ts]: The `delete*` docblocks need to be updated with the type renaming.

File: package.json (hunk 5 above)
Comment (jeanmurillo95) [reply package.json]: Ohh I didn’t know,  I will use it.

File: tests/Repository.spec.ts (hunk 2 above)
Comment (DanStory) [reply tests/Repository.spec.ts]: One of my concerns with lodash is that is 4x slower than rfdc and gets worse with document size. It is unfortunate that `rfdc` has an issue with the newer [bson ObjectId](https://github.com/mongodb/js-bson). Have we tried to investigate the issue and patch/fix it ourselves?

File: src/Repository.ts (hunk 13 above)
Comment (DanStory) [reply src/Repository.ts]: Separating out the callback to a `createCollection` is fine, however the refactor of creating the collection indices (IMO) loses readability and maintainability due to the "sub-recursion" and argument instance manipulation.

File: src/Repository.ts [hunk 19]
```diff
@@ -357,54 +382,65 @@ export class MongoRepository<DOC, DTO = DOC> {
...
-                reject(indexErr);
-              }
-            }
-          }
-        }
-
-        resolve(ourCollection);
-      });
+      let ourCollection;
+      const exist = (await db.listCollections({ name: this.options.name }).toArray())[0];
```
Comment (DanStory) [inline src/Repository.ts]: I don't think we are really gaining anything with checking for existence, when we still can have a race condition in the creation of the collection in a distributed deployment.

**Review by DanStory — CHANGES_REQUESTED** [review]

File: tests/Repository.spec.ts (hunk 2 above)
Comment (MatthewEppelsheimer) [reply tests/Repository.spec.ts]: I agree it's worth digging into this further to avoid introducing Lodash, with its 4x slower speed, if at all possible.

File: src/Repository.ts [hunk 20]
```diff
@@ -144,7 +145,7 @@ export class MongoRepository<DOC, DTO = DOC> {
     const eventResult = await this.invokeEvents(PRE_KEY, [RepoOperation.save, RepoOperation.create], document);
     const res = await collection.insertOne(eventResult);
 
-    let newDocument = this.toggleId(res.ops[0], false);
+    let newDocument = this.toggleId({ id: res.insertedId }, false);
```
Comment (DanStory) [inline src/Repository.ts:148]: This needs to be the whole document, not just the id.

**Review by DanStory — CHANGES_REQUESTED** [review]
There is unaddressed comments, make sure to unhide all comments on the [Conversation tab](https://github.com/swimlane/mongtype/pull/39)

image

File: tests/Repository.spec.ts (hunk 2 above)
Comment (DanStory) [reply tests/Repository.spec.ts]: Might even suggest we move towards a default shallow copy (due to `toggleId` alterations to the document) and the user/consumer of the lib brings their own clone, as in our usage the `_id` is the only `ObjectId` and can work around using `rfdc` still.

**Comment by MatthewEppelsheimer** [comment]
@jeanmurillo95 there appear to still be several unresolved comments from the last review.

**Comment by youngcm2** [comment]
@MatthewEppelsheimer @DanStory Can you please look at this PR?

File: package.json (hunk 3 above)
Comment (MatthewEppelsheimer) [reply package.json]: @jeanmurillo95 marked as resolved, but the package major version still needs to be updated.

Comment (MatthewEppelsheimer) [reply package.json]: @jeanmurillo95 I know various teams do this differently, but Automation Team prefers to have the person who started a PR comment thread be the one to eventually mark it as resolved. This make it easier to confirm resolution during re-review, and see outstanding issues at a glance.

File: package.json (hunk 8 above)
Comment (MatthewEppelsheimer) [reply package.json]: @jeanmurillo95 this still needs to be rolled back to `^16.x.x`.

File: package.json (hunk 3 above)
Comment (youngcm2) [reply package.json]: @DanStory @MatthewEppelsheimer version changed to 7.0.0

File: src/Repository.ts (hunk 9 above)
Comment (MatthewEppelsheimer) [reply src/Repository.ts:50]: @jeanmurillo95 this remains unaddressed. Do you understand the feedback? 

On line 88, the `_Id` is replaced with `Id`. So the change to the return type here and on L83 are incorrect. (`tsc` isn't catching this because `toggleId()` returns `any`; fixing that would be a big improvement, but I don't recommend trying to fix that right now.)

File: src/Repository.ts (hunk 11 above)
Comment (MatthewEppelsheimer) [reply src/Repository.ts:345]: Still unaddressed. Do you understand the issue?

Re-assigning a variable that was storing an object reference overwrites that variable's reference to the object in local scope, without affecting the value of the variable in the calling scope. Neither `newDocument` or `originalDocument` will be changed from the caller's perspective.

File: src/Repository.ts [hunk 21]
```diff
@@ -357,54 +392,51 @@ export class MongoRepository<DOC, DTO = DOC> {
...
-            if (createErr.codeName === 'NamespaceExists') {
-              // race condition. ignore for now, as I can't seem to get
-              // transactions to work in mongo 4.4 as yet
-              ourCollection = db.collection(this.options.name);
-            } else {
-              reject(err);
-            }
+      let ourCollection;
+      try {
+        ourCollection = await db.collection(this.options.name);
```
Comment (DanStory) [inline src/Repository.ts:397]: This is missing `{ strict: true }` so it doesn't create the collection out right

**Review by DanStory — CHANGES_REQUESTED** [review]

File: src/Repository.ts (hunk 9 above)
Comment (DanStory) [reply src/Repository.ts:50]: This is unresolved

File: src/Repository.ts (hunk 11 above)
Comment (DanStory) [reply src/Repository.ts:345]: This is unresolved

File: src/Repository.ts (hunk 20 above)
Comment (DanStory) [reply src/Repository.ts:148]: This is unresolved
